using_restricted_array.rb

```diff
@@ -7,41 +7,89 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  final_length = 0
+  while array[final_length] != nil
+    final_length += 1
+  end
+  return final_length
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  n = length(array)
+  n.times do |i|
+    print "#{array[i]} "
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  length.times do |i|
+    if array[i]==value_to_find
+      return true
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  max = array[0]
+  length.times do |i|
+    if array[i] > max
+      max = array[i]
+    end
+  end
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  min = array[0]
+  length.times do |i|
+    if array[i] < min
+      min = array[i]
+    end
+  end
+  return min
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  j = length - 1
+  while i < j
+    first = array[i]
+    last = array[j]
+    array[i] = last
+    array[j] = first
+    i += 1
+    j -= 1
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  low = 0
+  high = length-1
+  mid = (length/2.0).round()
+  while low <= high
```

Comment:
Please take a look at the updated binary_search slide. 
Here's one of the considerations I was missing in the algorithm on the slide before the update: Consider the scenario where there are two elements in the array and the value_to_find is the second element in the array. In the first iteration, min = 0, max=1, mid=0. Value is not at index 0. So, min = mid (still 0). On re-entering the loop, mid would be still 0. So we had an infinite loop. (To fix this, update max to be mid -1, if value_to_find is less than mid. And update min = mid+1, if value_to_find is greater than mid.)